### PR TITLE
Enhancement on stock.get_quote_history API

### LINKED ISSUE #128: 个别美股获取不了或者获取的数据有误
我在使用获取历史数据的接口函数efinance.stock.get_quote_history(stock_codes: Union[str, List[str]], beg: str = '19000101', end: str = '20500101', klt: int = 101, fqt: int = 1, **kwargs)时发现个别美股如MRK(默沙东)、PLD(安博)输入美股代码后无法获取到相应数据；个别股如PG（宝洁）输入美股代码获取到的是AAPL(苹果)的数据，以上三个代码好像只能通过输入中文字符来获取 请问有可能通过美股股票代码获取到这些股票的历史数据吗？

### PR DESCRIPTION
**Pull request by allen-li1231**
### This PR enhanced stock.get_quote_history API with the following aspects:
* Exact symbol match.
* Add MarketType enum for searching stock symbol on specified stock market.
* Expose toggling on cached EastMoney stock id.
* Enrich corresponding comments.

### Feature examples:

```python
import efinance as ef

ef.stock.get_quote_history("PG")
# unexpectedly returns AAPL's stock history

# get history of stock, whose symbol exactly matches "PG"
ef.stock.get_quote_history("PG", quote_symbol_mode=True, use_id_cache=False)  # do not use cache as it will still matches AAPL
# Correctly returns P&G stock history

# fuzzy search "PG" in A stock market
ef.stock.get_quote_history("PG", market_type=ef.utils.MarketType.A_stock, use_id_cache=False)
# returns 平高电气 stock history

# exact search "PG" in A stock market
ef.stock.get_quote_history("PG", market_type=ef.utils.MarketType.A_stock, quote_symbol_mode=True, use_id_cache=False)
# returns empty dataframe as none matches in reality

```

### CODE REVIEW

**Comment by allen-li1231** [comment]
@Micro-sheep Let me know if there is any advice or improvements as this is my first pr.

**Comment by Micro-sheep** [comment]
> @Micro-sheep Let me know if there is any advice or improvements as this is my first pr.

I will be reviewing the code in the next two week.

File: changelog.md [hunk 1]
```diff
@@ -1,5 +1,11 @@
 # Changelog
 
+## v0.5.0(2023-02-22)
```
Comment (Micro-sheep) [inline changelog.md:3]: ## v0.5.1(2024-02-22)

File: efinance/utils/__init__.py [hunk 2]
```diff
@@ -133,36 +156,55 @@ def search_quote(
...
+                and (market_type is None or (market_type.value == item['Classify']))
+        ]
         # NOTE 暂时仅存储第一个搜索结果
         save_search_result(keyword, quotes[:1])
         if count == 1:
-            return quotes[0]
-        return quotes
+            return quotes[0] if len(quotes) >= 1 else None
+
+        return quotes[:count + 1]
```
Comment (Micro-sheep) [inline efinance/utils/__init__.py:190]: It should be `return quotes[:count]`

File: efinance/stock/getter.py [hunk 3]
```diff
@@ -221,8 +237,8 @@ def get_quote_history(
     4759  贵州茅台  600519  2021-07-28  1703.00  1768.90  1788.20  1682.12   85369  1.479247e+10  6.19  3.27  56.01   0.68
     4760  贵州茅台  600519  2021-07-29  1810.01  1749.79  1823.00  1734.34   63864  1.129957e+10  5.01 -1.08 -19.11   0.51
 
-    >>> # 获取多只股票历史行情
-    >>> stock_df = ef.stock.get_quote_history(['600519','300750'])
+    >>> # 获取多只A股股票历史行情
```
Comment (Micro-sheep) [inline efinance/stock/getter.py:240]: Keep your new example and the example of obtaining market history without restricting market_type.

File: efinance/utils/__init__.py (hunk 2 above)
Comment (allen-li1231) [reply efinance/utils/__init__.py]: See [correct condition of multiple returns of quote info](https://github.com/Micro-sheep/efinance/pull/161/commits/3b96e43c3ab16f53cc7e4d9f718f47dd0928739e)

File: efinance/stock/getter.py (hunk 3 above)
Comment (allen-li1231) [reply efinance/stock/getter.py]: See [retrieve original example, while illustrate new proposed features](https://github.com/Micro-sheep/efinance/pull/161/commits/6e50963c9dafde40e001131246dc4774b502a95e)

**Comment by allen-li1231** [comment]
Corrected code as suggested.

File: efinance/stock/getter.py [hunk 4]
```diff
@@ -43,7 +43,7 @@
 # * 适配 pythn 3.10 及其以上版本
 if python_version >= (3, 10):
     requests.packages.urllib3.util.ssl_.DEFAULT_CIPHERS = 'ALL:@SECLEVEL=1'
-
+    requests.packages.urllib3.disable_warnings()
```
Comment (Micro-sheep) [inline efinance/stock/getter.py:46]: This is also for python version bellow 3.10.

File: efinance/common/config.py [hunk 5]
```diff
@@ -1,11 +1,24 @@
...
 
 HERE = Path(__file__).parent
 
 
+class MarketType(enum.Enum):
+    A_stock = "AStock"
+    Hongkong = "HK"
+    London_stock_exchange = "LSE"
+    US_stock = "UsStock"
+    # there may exists more market type
```
Comment (Micro-sheep) [inline efinance/common/config.py:12]: Could you please add more market types? You can get more  `classify` by `efinance.utils.search_quote` with stock and stock index from different markets.

For example
```python
In [1]: from efinance.utils import search_quote

In [2]: search_quote('000001',count=2)
Out[2]:
[Quote(code='000001', name='平安银行', pinyin='PAYH', id='0000012', jys='6', classify='AStock', market_type='2', security_typeName='深A', security_type='2', mkt_num='0', type_us='6', quote_id='0.000001', unified_code='000001', inner_code='15855238340410'),
 Quote(code='000001', name='上证指数', pinyin='SZZS', id='0000011', jys='1', classify='Index', market_type='1', security_typeName='指数', security_type='5', mkt_num='1', type_us='1', quote_id='1.000001', unified_code='000001', inner_code='38709193299776')]

In [3]: search_quote('NDX',count=2)
Out[3]:
[Quote(code='NDX', name='纳斯达克', pinyin='NSDK', id='NDX_UI', jys='QQZS', classify='UniversalIndex', market_type='_UI', security_typeName='指数', security_type='11', mkt_num='100', type_us='0', quote_id='100.NDX', unified_code='NDX', inner_code='49387883367286'),
 Quote(code='601198', name='东兴证券', pinyin='DXZQ', id='6011981', jys='2', classify='AStock', market_type='1', security_typeName='沪A', security_type='1', mkt_num='1', type_us='2', quote_id='1.601198', unified_code='601198', inner_code='32322054865723')]
```

Comment (allen-li1231) [reply efinance/common/config.py:12]: Below lists classify names I have discovered so far:
```python
class MarketType(enum.Enum):
    A_stock = "AStock"  # A股
    A_stock_index = "Index" # A股指数
    B_stock = "BStock"  # B股
    index = "Index" # 沪深京指数
    STAR_market = "23"  # 科创板
    CSI_free_float = "24" # 中证系列指数
    NEEQ = "NEEQ"   # 京A，新三板（全国中小企业股份转让系统）
    BK = "BK"   # 板块
    Hongkong = "HK" # 港股
    US_stock = "UsStock"    # 美股
    London_stock = "LSE"   # 英股
    London_stock_international = "LSEIOB"   # 伦敦交易所国际挂盘册
    universal_index = "UniversalIndex"  # 国外指数
    SIX_Swiss = "SIX" # SIX瑞士股市
```
I think the granularity of market type might be too fined for users. For example, when users want to search for indices in China, they probably expect API to search from both "A_stock_index", "CSI_free_float" and "index", what do you think?

**Pull request merged**